### ISSUE TITLE
[REVIEW]: Underworld2: Python Geodynamics Modelling for Desktop, HPC and Cloud

### ISSUE THREAD

**Issue by whedon** [openjournals/joss-reviews#1797]
**Submitting author:** @jmansour (<a href="http://orcid.org/0000-0001-5865-1664">john mansour</a>)
**Repository:** <a href="https://github.com/underworldcode/underworld2" target ="_blank">https://github.com/underworldcode/underworld2</a>
**Version:** v2.9.2b
**Editor:** @leouieda
**Reviewer:** @gassmoeller, @gabersyd
**Archive:** <a href="https://doi.org/10.5281/zenodo.3687399" target="_blank">10.5281/zenodo.3687399</a>

## Status

[![status](https://joss.theoj.org/papers/4e96f9dba4f8f129f64770e74abdec0e/status.svg)](https://joss.theoj.org/papers/4e96f9dba4f8f129f64770e74abdec0e)

Status badge code:

```
HTML: <a href="https://joss.theoj.org/papers/4e96f9dba4f8f129f64770e74abdec0e"><img src="https://joss.theoj.org/papers/4e96f9dba4f8f129f64770e74abdec0e/status.svg"></a>
Markdown: [![status](https://joss.theoj.org/papers/4e96f9dba4f8f129f64770e74abdec0e/status.svg)](https://joss.theoj.org/papers/4e96f9dba4f8f129f64770e74abdec0e)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="https://github.com/underworldcode/underworld2" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) by leaving comments in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

## Reviewer instructions & questions

@gassmoeller & @gabersyd, please carry out your review in this issue by updating the checklist below. If you cannot edit the checklist please:
1. Make sure you're logged in to your GitHub account
2. Be sure to accept the invite at this URL: https://github.com/openjournals/joss-reviews/invitations

The reviewer guidelines are available here: https://joss.readthedocs.io/en/latest/reviewer_guidelines.html. Any questions/concerns please let @leouieda know.

✨ **Please try and complete your review in the next two weeks** ✨ 

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/underworldcode/underworld2">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@jmansour) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/underworldcode/underworld2">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@jmansour) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

**Comment by whedon** [comment]
Hello human, I'm @whedon, a robot that can help you with some common editorial tasks. @gassmoeller, @gabersyd it looks like you're currently assigned to review this paper :tada:.

**:star: Important :star:**

If you haven't already, you should seriously consider unsubscribing from GitHub notifications for this (https://github.com/openjournals/joss-reviews) repository. As a reviewer, you're probably currently watching this repository which means for GitHub's default behaviour you will receive notifications (emails) for all reviews 😿

To fix this do the following two things:

1. Set yourself as 'Not watching' https://github.com/openjournals/joss-reviews:

![watching](https://cloud.githubusercontent.com/assets/4483/20250593/64d7ce48-a9de-11e6-9225-d3dfb3e48e68.png)

2. You may also like to change your default settings for this watching repositories in your GitHub profile here: https://github.com/settings/notifications

![notifications](https://cloud.githubusercontent.com/assets/4483/20250604/79eb3144-a9de-11e6-9e51-0fe940966689.png)


For a list of things I can do to help you, just type:

```
@whedon commands
```

For example, to regenerate the paper pdf after making changes in the paper's md or bib files, type:

```
@whedon generate pdf
```

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
PDF failed to compile for issue #1797 with the following error: 

 Can't find any papers to compile :-(

**Comment by leouieda** [comment]
@whedon generate pdf from branch joss

**Comment by whedon** [comment]
```
Attempting PDF compilation from custom branch joss. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01797/joss.01797/10.21105.joss.01797.pdf)

**Comment by leouieda** [comment]
👋 @jmansour @gassmoeller @gabersyd this is the review thread for the paper. All of our communications will happen here from now on.

Both reviewers have checklists at the top of this thread with the JOSS requirements. As you go over the submission, please check any items that you feel have been satisfied. There are also links to the JOSS reviewer guidelines.

The JOSS review is different from most other journals. Our goal is to work with the authors to help them meet our criteria instead of merely passing judgment on the submission. As such, the reviewers are encouraged to submit issues and pull requests on the software repository. When doing so, please mention `openjournals/joss-reviews#1797` so that a link is created to this thread (and I can keep an eye on what is happening). Please also feel free to comment and ask questions on this thread. In my experience, it is better to post comments/questions/suggestions as you come across them instead of waiting until you've reviewed the entire package.

We aim for reviews to be completed within about 4 weeks. Please let me know if any of you require some more time. We can also use Whedon (our bot) to set automatic reminders if you know you'll be away for a known period of time.

Please feel free to ping me here (@leouieda) or email me privately if you have any questions/concerns.

**Comment by gassmoeller** [comment]
@leouieda: I have a question about the following item on the reviewer checklist:

```
State of the field: Do the authors describe how this software compares to other commonly-used packages?
```

The authors do not explicitly describe this in the paper (although they do describe their main design principle of an easily usable library, which differs somewhat from many of the other software packages in our field), but I also have two problems with this item. First it is not mentioned on the https://joss.readthedocs.io/en/latest/submitting.html#submitting-a-paper-to-joss website, and neither on the https://joss.readthedocs.io/en/latest/review_criteria.html#review-criteria website. Thus, the authors were not asked to provide such a comparison. Second, this item can be a can of worms if you want to describe it thoroughly. I think this item should be revised somewhat to either only require a short / general comparison (including mentioning this expectation on the submission website), or it should be dropped from the review checklist. Do you think this is reasonable?

**Comment by jmansour** [comment]
@whedon generate pdf from branch joss

**Comment by whedon** [comment]
```
Attempting PDF compilation from custom branch joss
. Reticulating splines etc...
```

**Comment by whedon** [comment]
PDF failed to compile for issue #1797 with the following error: 

 error: pathspec 'joss
' did not match any file(s) known to git.
Can't find any papers to compile :-(

**Comment by jmansour** [comment]
@whedon generate pdf from branch joss

**Comment by whedon** [comment]
```
Attempting PDF compilation from custom branch joss. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01797/joss.01797/10.21105.joss.01797.pdf)

**Comment by leouieda** [comment]
> I think this item should be revised somewhat to either only require a short / general comparison (including mentioning this expectation on the submission website), or it should be dropped from the review checklist. Do you think this is reasonable?

Hi @gassmoeller that is probably a new item on the checklist. I hadn't seen it before. I'll bring this up with the JOSS editors and see what we can do about it. I agree that we should be very explicit about what we want from that statement. I imagine the idea was to provide a general comparison statement, not a thorough performance/accuracy/features comparison. Thanks for pointing this out!

**Comment by leouieda** [comment]
On that note, @gassmoeller @gabersyd what is the status of the review? Any updates?

**Comment by gassmoeller** [comment]
I am satisfied with the current state of the software and paper, and think this is an important contribution. The authors have responded productively and fixed the small comments and objections I had. I am only waiting for underworldcode/underworld2#428 to be fixed (one of the authors is currently not on the paper, because he has no ORCID so far), and afterwards this is ready from my side. I particularly liked the easy distribution and online tutorials. Nice work!

**Comment by gabersyd** [comment]
@leouieda, I need one more week to give a complete feedback.

Gabriele Morra
<http://scholar.google.com/citations?sortby=pubdate&hl=en&user=DDuTppAAAAAJ&view_op=list_works>
, Department of Physics <http://physics.louisiana.edu>, School of
Geosciences <http://geos.louisiana.edu>, University of Louisiana at
Lafayette <http://www.louisiana.edu>

**Comment by jmansour** [comment]
Thanks @gassmoeller & @gabersyd

**Comment by jmansour** [comment]
Hi @leouieda 

Are the ORCID identifiers strictly necessary for submission?  One of our authors is somewhat privacy concerned and would rather not sign up for an ORCID.

**Comment by leouieda** [comment]
@jmansour ORCIDs are encouraged but not strictly required if an author feels strongly about not providing one (our own [example paper](https://joss.readthedocs.io/en/latest/submitting.html#example-paper-and-bibliography) has an authors without an ORCID).

**Comment by leouieda** [comment]
@gassmoeller thank you for your review! :clap: 

@gabersyd no worries, take your time :+1:

**Comment by jmansour** [comment]
Thanks @leouieda 

A further question if I may.  One author has since moved onto other employment and is no longer directly involved in the project.  I assume that their affiliation should represent their place of employment at the time of their contribution, but possibly their current employment should be also  included as means of contact.  Does JOSS have any particular preference in this situation?

**Comment by leouieda** [comment]
@jmansour not that I'm aware of. You could always add the new affiliation and list the author as having both.

**Comment by jmansour** [comment]
@whedon generate pdf from branch joss

**Comment by whedon** [comment]
```
Attempting PDF compilation from custom branch joss. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01797/joss.01797/10.21105.joss.01797.pdf)

**Comment by gabersyd** [comment]
@whedon commands

**Comment by whedon** [comment]
Here are some things you can ask me to do:

```
# List Whedon's capabilities
@whedon commands

# List of editor GitHub usernames
@whedon list editors

# List of reviewers together with programming language preferences and domain expertise
@whedon list reviewers

EDITORIAL TASKS

# Compile the paper
@whedon generate pdf

# Compile the paper from alternative branch
@whedon generate pdf from branch custom-branch-name

# Ask Whedon to check the references for missing DOIs
@whedon check references

```

**Comment by danielskatz** [comment]
👋 @gabersyd - How is this going?

31 October:
> @leouieda, I need one more week to give a complete feedback.

**Comment by gabersyd** [comment]
A comment: the installation folder (underworld2/docs/install_guides/) contains examples only for large HPC facilities. It would be useful to offer one also for a smaller machine, so that students could try using it in a smaller machine like a powerful desktop or a small cluster that they might want to access. 

I tried to follow some existing instructions like the one for Caltech fram on our Queen Bee cluster in Louisiana, but I was stuck as the system asked me to install "scons". I did not see this requirement in the instructions.

**Comment by gabersyd** [comment]
I finished the review. I just have an issue as I struggled with
dependencies when trying to install it on our largest cluster here in
Louisiana. Maybe we can look at it together, as the dependencies were not
all clear to me when I tried.

Thank you
Gabriele

Gabriele Morra
<http://scholar.google.com/citations?sortby=pubdate&hl=en&user=DDuTppAAAAAJ&view_op=list_works>
Associate Professor <https://physics.louisiana.edu/about-us/faculty-staff>
Department of Physics <http://physics.louisiana.edu> and School of
Geosciences <http://geos.louisiana.edu>
University of Louisiana at Lafayette <http://www.louisiana.edu>
Editor in Chief of *Artificial Intelligence in Geosciences
<http://www.keaipublishing.com/en/journals/artificial-intelligence-in-geosciences/>*

**Comment by jmansour** [comment]
Thanks @gabersyd! 

Note that there are updated installation instructions on the development (and joss) branch. Apologies, I thought I'd mentioned that earlier, but checking this thread I can't see mention of it. Here's the corresponding ReadTheDocs page:

https://underworld2.readthedocs.io/en/development/Installation.html

I think that should give reasonable coverage for native machine installation. More than happy to have a closer look at installation on Queen Bee so see where things are going awry.

**Comment by gabersyd** [comment]
Hi John, thank you for your answer. I have tried to follow all the
instructions and got stuck again into the installation of scons. I am not
familiar with this library. I looker at scons.org and it seems that they
have mostly developed for
Python 2.*, which is quite obsolete. I tried to install it manually on the
cluster, and I have gotten:

UnsatisfiableError: The following specifications were found to be in
conflict:

  - python 3.5*

  - scons -> python 2.7*

Use "conda info <package>" to see the dependencies for each package.


Is scons so necessary for the installation of Underworld, that there is no
way to get around it? Very few Python users today use Python 2.*, now that
it arrive to Python 3.8.

Do you suggest me just to use Docker? I am concerned whether the
performance on the cluster will be the same if I use Docker instead of a
native installation. Is it just the same?

Gabriele Morra
<http://scholar.google.com/citations?sortby=pubdate&hl=en&user=DDuTppAAAAAJ&view_op=list_works>
Associate Professor <https://physics.louisiana.edu/about-us/faculty-staff>
Department of Physics <http://physics.louisiana.edu> and School of
Geosciences <http://geos.louisiana.edu>
University of Louisiana at Lafayette <http://www.louisiana.edu>
Editor in Chief of *Artificial Intelligence in Geosciences
<http://www.keaipublishing.com/en/journals/artificial-intelligence-in-geosciences/>*

**Comment by jmansour** [comment]
Unfortunately `scons` is the package we use to build Underworld, so it is necessary.  There should be no issue with using it with Py3, but I don't have experience with `conda`.  Typically, I just use `pip install scons` (or perhaps `pip install --user scons` for local install).   Actually, you should simply be able to download it from the `scons` project page, untar, and then update your `PATH` environment variable so that it can locate the `scons` Python executable (in the untarred `script` directory). 

https://scons.org/pages/download.html

Docker is low (Windows/OSX) or no (Linux) overhead. However, on cluster machines, there are caveats to usage, and it is generally better to use something like Singularity or Shifter which act as conduits to Docker environments in HPC environments.  It really depends on how the machine is being used and you might need to chat with the system admins to see what is the best way forward. Actually there are even performance advantages to using Docker on HPC for very large Python jobs, however, if not already setup, native build will probably be the path of least resistance. I'm happy to help if I can in either case.

**Comment by gabersyd** [comment]
Thank you, John, I think that my review is finished. I am sure that many will use this version of Underworld as it is more user friendly than the previous version.

**Comment by leouieda** [comment]
@gabersyd thank you for your review! I see that all review items have been checked so I assume that you are happy to sign off on this.

SCons used to be Python 2.7 only but recent versions are Python 3.5+ only. Installing from PyPI (using pip) will get you the latest version (3.1.2), which should work. You can also install with conda from [conda-forge](https://github.com/conda-forge/scons-feedstock) where the latest version is 3.1.2 and should work in Python 3 (`conda install scons=3.1.2 -c conda-forge`). The issue is probably that the Anaconda package for SCons is outdated.

**Comment by leouieda** [comment]
@whedon check referencess

**Comment by leouieda** [comment]
@whedon check references

**Comment by leouieda** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
PDF failed to compile for issue #1797 with the following error: 

 Can't find any papers to compile :-(

**Comment by leouieda** [comment]
@whedon generate pdf from branch joss

**Comment by whedon** [comment]
```
Attempting PDF compilation from custom branch joss. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01797/joss.01797/10.21105.joss.01797.pdf)

**Comment by whedon** [comment]
```
Reference check summary:

OK DOIs

- 10.21105/joss.01136 is OK
- 10.5281/zenodo.3348451 is OK
- 10.1016/j.pepi.2018.07.009 is OK
- 10.5194/se-10-969-2019 is OK

MISSING DOIs

- https://doi.org/10.1007/978-3-0348-8197-5_10 may be missing for title: Mantle Convection Modeling with Viscoelastic/Brittle Lithosphere: Numerical Methodology and Plate Tectonic Modeling
- https://doi.org/10.1016/s0021-9991(02)00031-1 may be missing for title: A Lagrangian integration point finite element method for large deformation modeling of viscoelastic geomaterials
- https://doi.org/10.1016/j.pepi.2007.06.009 may be missing for title: Computational approaches to studying non-linear dynamics of the crust and mantle 
- https://doi.org/10.1130/g39943.1 may be missing for title: Formation of cratonic lithosphere during the initiation of plate tectonics

INVALID DOIs

- None
```

**Comment by whedon** [comment]
```
Reference check summary:

OK DOIs

- 10.21105/joss.01136 is OK
- 10.5281/zenodo.3348451 is OK
- 10.1016/j.pepi.2018.07.009 is OK
- 10.5194/se-10-969-2019 is OK

MISSING DOIs

- https://doi.org/10.1007/978-3-0348-8197-5_10 may be missing for title: Mantle Convection Modeling with Viscoelastic/Brittle Lithosphere: Numerical Methodology and Plate Tectonic Modeling
- https://doi.org/10.1016/s0021-9991(02)00031-1 may be missing for title: A Lagrangian integration point finite element method for large deformation modeling of viscoelastic geomaterials
- https://doi.org/10.1016/j.pepi.2007.06.009 may be missing for title: Computational approaches to studying non-linear dynamics of the crust and mantle 
- https://doi.org/10.1130/g39943.1 may be missing for title: Formation of cratonic lithosphere during the initiation of plate tectonics

INVALID DOIs

- None
```

**Comment by leouieda** [comment]
@jmansour could you please add the missing DOIs for the references outlined by Whedon above?

**Comment by jmansour** [comment]
@whedon generate pdf from branch joss

**Comment by whedon** [comment]
```
Attempting PDF compilation from custom branch joss. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01797/joss.01797/10.21105.joss.01797.pdf)

**Comment by jmansour** [comment]
Ok, I think they should all be good now @leouieda.  

And can confirm using `scons 3.12` with Py3 works fine. Thanks for input @leouieda.

**Comment by leouieda** [comment]
@jmansour thank you for the update. I submitted a PR with some editorial fixes (mostly references and citation style). Let me know once you review and merge these changes.

**Comment by jmansour** [comment]
@whedon generate pdf from branch joss

**Comment by whedon** [comment]
```
Attempting PDF compilation from custom branch joss. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01797/joss.01797/10.21105.joss.01797.pdf)

**Comment by jmansour** [comment]
@whedon generate pdf from branch joss

**Comment by whedon** [comment]
```
Attempting PDF compilation from custom branch joss
. Reticulating splines etc...
```

**Comment by whedon** [comment]
PDF failed to compile for issue #1797 with the following error: 

 error: pathspec 'joss
' did not match any file(s) known to git.
Can't find any papers to compile :-(

**Comment by jmansour** [comment]
@whedon generate pdf from branch joss

**Comment by whedon** [comment]
```
Attempting PDF compilation from custom branch joss. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01797/joss.01797/10.21105.joss.01797.pdf)

**Comment by jmansour** [comment]
Thanks @leouieda, I've merged in those changes.  Also I've added some further minor formatting tweaks:
https://github.com/underworldcode/underworld2/commit/15052cf73429bedde7b564d5582b681f6a64507b
https://github.com/underworldcode/underworld2/commit/e2310ee64c57bbc30070810374d858e3bc075f3e
https://github.com/underworldcode/underworld2/commit/66057689d27f6870b0b446a8b44590efc3f0b2b3

**Comment by leouieda** [comment]
@jmansour I'm happy to move on with acceptance of your paper :tada: Here are the steps you need to take now:

- Double check authors and affiliations (including ORCIDs)
- Make a release with the latest changes from the review and **post the version number here**.
- Archive the release on Zenodo/figshare/etc and **post the DOI here**.
- Make sure that the title and author list (including ORCIDs) in the archive match the JOSS paper.

**Comment by jmansour** [comment]
Great, thanks @leouieda. Will report back shortly.

**Comment by leouieda** [comment]
:wave: hi @jmansour, just checking in on the progress here.

**Comment by jmansour** [comment]
Hi @leouieda 

Just finalising some details for the release.  Should be ready next week.

**Comment by jmansour** [comment]
Hi @leouieda 

Ok, we have a new release out. Version 2.9.0b:

https://github.com/underworldcode/underworld2/releases/tag/v2.9.0b

Still waiting on Zenodo to generate the new corresponding DOI, but it should show up shortly here:
https://doi.org/10.5281/zenodo.1436039

Affiliations and ORCIDs have been confirmed. Note that as discussed earlier, one author has requested not to have an associated ORCID.   

Let me know if you need anything further. 

Thanks

**Comment by jmansour** [comment]
Ok, seems there was an issue with our `.zenodo.json` .  Fixed now.  

DOI:  10.5281/zenodo.3661465

https://zenodo.org/record/3661465

**Comment by jmansour** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01797/joss.01797/10.21105.joss.01797.pdf)

**Comment by leouieda** [comment]
Thank you, I'll have a look at it shortly.

**Comment by leouieda** [comment]
@jmansour apologies for the delay. Thank you for posting the DOI link and version number. I double checked the paper and everything looks good. The only minor thing I have to ask is that you update the title of the Zenodo archive of https://zenodo.org/record/3661465 to match the JOSS paper title. Let me know when you get a chance to do that and we'll move on to acceptance :tada:

**Comment by leouieda** [comment]
@whedon set 2.9.0b as version

**Comment by whedon** [comment]
OK. 2.9.0b is the version.

**Comment by leouieda** [comment]
@whedon set 10.5281/zenodo.3661465 as archive

**Comment by whedon** [comment]
OK. <a href="https://doi.org/10.5281/zenodo.3661465" target="_blank">10.5281/zenodo.3661465</a> is the archive.

**Comment by leouieda** [comment]
@whedon accept

**Comment by whedon** [comment]
```
Attempting dry run of processing paper acceptance...
```

**Comment by whedon** [comment]
Check final proof :point_right: https://github.com/openjournals/joss-papers/pull/1341

If the paper PDF and Crossref deposit XML look good in https://github.com/openjournals/joss-papers/pull/1341, then you can now move forward with accepting the submission by compiling again with the flag `deposit=true` e.g.
 ```
@whedon accept deposit=true
```

**Comment by jmansour** [comment]
Hi @leouieda 

Am I able to update the title of an existing Zenodo archive?   In any case, I've created a new release (v2.9.2) with the updated title. Here's the corresponding Zenodo reference: 
https://zenodo.org/record/3687399

**Comment by leouieda** [comment]
@jmansour thanks for the update

**Comment by leouieda** [comment]
@whedon set v2.9.2b as version

**Comment by whedon** [comment]
OK. v2.9.2b is the version.

**Comment by leouieda** [comment]
@whedon set 10.5281/zenodo.3687399 as archive

**Comment by whedon** [comment]
OK. <a href="https://doi.org/10.5281/zenodo.3687399" target="_blank">10.5281/zenodo.3687399</a> is the archive.

**Comment by leouieda** [comment]
@whedon accept

**Comment by whedon** [comment]
```
Attempting dry run of processing paper acceptance...
```

**Comment by whedon** [comment]
:wave: @openjournals/joss-eics, this paper is ready to be accepted and published

. Check final proof :point_right: https://github.com/openjournals/joss-papers/pull/1350

If the paper PDF and Crossref deposit XML look good in https://github.com/openjournals/joss-papers/pull/1350, then you can now move forward with accepting the submission by compiling again with the flag `deposit=true` e.g.
 ```
@whedon accept deposit=true
```

**Comment by leouieda** [comment]
:wave: Hi @@openjournals/joss-eics just making sure this didn't get lost in the current updates to Whedon.

**Comment by Kevin-Mattheus-Moerman** [comment]
@leouieda Thanks. I'm on it.

**Comment by Kevin-Mattheus-Moerman** [comment]
I read the paper and think it looks good.

**Comment by Kevin-Mattheus-Moerman** [comment]
@whedon accept deposit=true

**Comment by whedon** [comment]
```
Doing it live! Attempting automated processing of paper acceptance...
```

**Comment by whedon** [comment]
🐦🐦🐦 👉 [Tweet for this paper](https://twitter.com/JOSS_TheOJ/status/1235895879233687555) 👈 🐦🐦🐦

**Comment by whedon** [comment]
🚨🚨🚨 **THIS IS NOT A DRILL, YOU HAVE JUST ACCEPTED A PAPER INTO JOSS!** 🚨🚨🚨

 Here's what you must now do:

0. Check final PDF and Crossref metadata that was deposited :point_right: https://github.com/openjournals/joss-papers/pull/1354
1. Wait a couple of minutes to verify that the paper DOI resolves [https://doi.org/10.21105/joss.01797](https://doi.org/10.21105/joss.01797)
2. If everything looks good, then close this review issue.
3. Party like you just published a paper! 🎉🌈🦄💃👻🤘

 Any issues? notify your editorial technical team...

**Comment by whedon** [comment]
:tada::tada::tada: Congratulations on your paper acceptance! :tada::tada::tada:

If you would like to include a link to your paper from your README use the following code snippets:

```
Markdown:
[![DOI](https://joss.theoj.org/papers/10.21105/joss.01797/status.svg)](https://doi.org/10.21105/joss.01797)

HTML:
<a style="border-width:0" href="https://doi.org/10.21105/joss.01797">
  <img src="https://joss.theoj.org/papers/10.21105/joss.01797/status.svg" alt="DOI badge" >
</a>

reStructuredText:
.. image:: https://joss.theoj.org/papers/10.21105/joss.01797/status.svg
   :target: https://doi.org/10.21105/joss.01797
```
This is how it will look in your documentation:

[![DOI](https://joss.theoj.org/papers/10.21105/joss.01797/status.svg)](https://doi.org/10.21105/joss.01797)

**We need your help!**

Journal of Open Source Software is a community-run journal and relies upon volunteer effort. If you'd like to support us please consider doing either one (or both) of the the following:

- Volunteering to review for us sometime in the future. You can add your name to the reviewer list here: https://joss.theoj.org/reviewer-signup.html
- Making a small donation to support our running costs here: https://numfocus.org/donate-to-joss

**Comment by leouieda** [comment]
Thanks @Kevin-Mattheus-Moerman, didn't mean to pressure.

@jmansour congratulations on your publication and I hope the JOSS submission/review process was enjoyable :tada: :confetti_ball: 

@gassmoeller @gabersyd thank you very for your reviews and feedback :1st_place_medal:

**Comment by jmansour** [comment]
Excellent result!  

I'd also like to thank @gassmoeller & @gabersyd for the time they put in reviewing this and the valuable feedback they provided. 🙏🙏

And thanks to @leouieda & @Kevin-Mattheus-Moerman and the rest of the JOSS crew. The review process was straightforward and even fun. 🙂
